In [1]:
import pandas as pd
import numpy as np

In [8]:
def load_dataset(): 
    sales = pd.read_excel("sales.xlsx",header =1)
    view_rate = pd.read_excel("view_rate.xlsx",header =1, index_col= 0)
    return sales, view_rate

sales = load_dataset()[0]
view_rate = load_dataset()[1]

                 방송일시  노출(분)    마더코드    상품코드            상품명 상품군   판매단가  \
0 2019-01-01 06:00:00   20.0  100346  201072  테이트 남성 셀린니트3종  의류  39900   
1 2019-01-01 06:00:00    NaN  100346  201079  테이트 여성 셀린니트3종  의류  39900   
2 2019-01-01 06:20:00   20.0  100346  201072  테이트 남성 셀린니트3종  의류  39900   
3 2019-01-01 06:20:00    NaN  100346  201079  테이트 여성 셀린니트3종  의류  39900   
4 2019-01-01 06:40:00   20.0  100346  201072  테이트 남성 셀린니트3종  의류  39900   

         취급액  
0  2099000.0  
1  4371000.0  
2  3262000.0  
3  6955000.0  
4  6672000.0  
       2019-01-01  2019-01-02  2019-01-03  2019-01-04  2019-01-05  2019-01-06  \
시간대                                                                             
02:00         0.0         0.0       0.000         0.0         0.0         0.0   
02:01         0.0         0.0       0.000         0.0         0.0         0.0   
02:02         0.0         0.0       0.000         0.0         0.0         0.0   
02:03         0.0         0.0       0.014         0.0       

In [9]:
sales.head()

방송일시  노출(분)    마더코드    상품코드            상품명 상품군   판매단가  \
0 2019-01-01 06:00:00   20.0  100346  201072  테이트 남성 셀린니트3종  의류  39900   
1 2019-01-01 06:00:00    NaN  100346  201079  테이트 여성 셀린니트3종  의류  39900   
2 2019-01-01 06:20:00   20.0  100346  201072  테이트 남성 셀린니트3종  의류  39900   
3 2019-01-01 06:20:00    NaN  100346  201079  테이트 여성 셀린니트3종  의류  39900   
4 2019-01-01 06:40:00   20.0  100346  201072  테이트 남성 셀린니트3종  의류  39900   

         취급액  
0  2099000.0  
1  4371000.0  
2  3262000.0  
3  6955000.0  
4  6672000.0

In [10]:
view_rate.head()

2019-01-01  2019-01-02  2019-01-03  2019-01-04  2019-01-05  2019-01-06  \
시간대                                                                             
02:00         0.0         0.0       0.000         0.0         0.0         0.0   
02:01         0.0         0.0       0.000         0.0         0.0         0.0   
02:02         0.0         0.0       0.000         0.0         0.0         0.0   
02:03         0.0         0.0       0.014         0.0         0.0         0.0   
02:04         0.0         0.0       0.000         0.0         0.0         0.0   

       2019-01-07  2019-01-08  2019-01-09  2019-01-10  ...  2019-12-23  \
시간대                                                    ...               
02:00         0.0         0.0         0.0       0.000  ...         0.0   
02:01         0.0         0.0         0.0       0.005  ...         0.0   
02:02         0.0         0.0         0.0       0.005  ...         0.0   
02:03         0.0         0.0         0.0       0.005  ...         0.0   
02:04         0.0         0.0         0.0       0.005  ...         0.0   

       2019-12-24  2019-12-25  2019-12-26  2019-12-27  2019-12-28  2019-12-29  \
시간대                                                                             
02:00         0.0       0.000         0.0         0.0         0.0         0.0   
02:01         0.0       0.000         0.0         0.0         0.0         0.0   
02:02         0.0       0.000         0.0         0.0         0.0         0.0   
02:03         0.0       0.017         0.0         0.0         0.0         0.0   
02:04         0.0       0.017         0.0         0.0         0.0         0.0   

       2019-12-30  2019-12-31  2019-01-01 to 2019-12-31  
시간대                                                      
02:00         0.0       0.000                     0.003  
02:01         0.0       0.012                     0.003  
02:02         0.0       0.000                     0.004  
02:03         0.0       0.000                     0.004  
02:04         0.0       0.000                     0.004  

[5 rows x 366 columns]

In [11]:
# 무형 지우기
sales.drop(sales[sales["상품군"]=="무형"].index, inplace=True)

In [12]:
# 결측된 노출분 채우기
for i,row in sales.iterrows():
    val = row["노출(분)"]
    if i == 0 :
        continue
        
    if  np.isnan(val):
        if sales.loc[i,"방송일시"] == sales.loc[i-1,"방송일시"]:
            sales.loc[i,"노출(분)"] = sales.loc[i-1,"노출(분)"]
        else : 
            continue
            
sales.isnull().sum()

방송일시     0
노출(분)    0
마더코드     0
상품코드     0
상품명      0
상품군      0
판매단가     0
취급액      0
dtype: int64

In [60]:
# 마지막 열의 값이 mena일까? 
test_lst= np.round(np.mean(view_rate.iloc[:,:-1].values,axis=1),3)==view_rate.iloc[:,-1].values
print(test_lst.all())
for i in np.argwhere(test_lst ==False):
    print("{}번째 열들의 평균은 ".format(i))
    print("마지막 열 직전까지의 값 {}".format(np.mean(view_rate.iloc[i,:-1].values)))
    print("실제값 : {}\n".format(view_rate.iloc[i,-1].values))

False
[93]번째 열들의 평균은 
마지막 열 직전까지의 값 0.002495890410958904
실제값 : [0.003]

[120]번째 열들의 평균은 
마지막 열 직전까지의 값 0.0015013698630136987
실제값 : [0.001]

[628]번째 열들의 평균은 
마지막 열 직전까지의 값 0.004512328767123287
실제값 : [0.004]

[652]번째 열들의 평균은 
마지막 열 직전까지의 값 0.004506849315068493
실제값 : [0.004]

[664]번째 열들의 평균은 
마지막 열 직전까지의 값 0.004506849315068493
실제값 : [0.004]

[745]번째 열들의 평균은 
마지막 열 직전까지의 값 0.006504109589041095
실제값 : [0.006]

[785]번째 열들의 평균은 
마지막 열 직전까지의 값 0.006504109589041096
실제값 : [0.006]

[837]번째 열들의 평균은 
마지막 열 직전까지의 값 0.006504109589041096
실제값 : [0.006]

[962]번째 열들의 평균은 
마지막 열 직전까지의 값 0.006504109589041096
실제값 : [0.006]

[988]번째 열들의 평균은 
마지막 열 직전까지의 값 0.004498630136986301
실제값 : [0.005]

[1206]번째 열들의 평균은 
마지막 열 직전까지의 값 0.0034958904109589043
실제값 : [0.004]



해당 위치에서 mean 값이랑 값이 맞지 않습니다.

In [77]:
# 마지막 열의 값이 mena일까? 
test_lst= np.round(np.mean(view_rate.iloc[:-1,:].values,axis=0),3)==view_rate.iloc[-1,:].values
print(test_lst.all())
for i in np.argwhere(test_lst ==False):
    print("{}번째 열들의 평균은 ".format(i))
    print("마지막 열 직전까지의 값 {}".format(np.mean(view_rate.iloc[:-1,i].values)))
    print("실제값 : {}\n".format(view_rate.iloc[-1,i].values[0]))

False
[50]번째 열들의 평균은 
마지막 열 직전까지의 값 0.0044777777777777784
실제값 : 0.005

[57]번째 열들의 평균은 
마지막 열 직전까지의 값 0.003486111111111111
실제값 : 0.004

[110]번째 열들의 평균은 
마지막 열 직전까지의 값 0.0024874999999999997
실제값 : 0.003

[184]번째 열들의 평균은 
마지막 열 직전까지의 값 0.003491666666666667
실제값 : 0.004

[204]번째 열들의 평균은 
마지막 열 직전까지의 값 0.00851875
실제값 : 0.008

[230]번째 열들의 평균은 
마지막 열 직전까지의 값 0.009490972222222224
실제값 : 0.01

[263]번째 열들의 평균은 
마지막 열 직전까지의 값 0.005501388888888889
실제값 : 0.005

[310]번째 열들의 평균은 
마지막 열 직전까지의 값 0.004525694444444444
실제값 : 0.004

[353]번째 열들의 평균은 
마지막 열 직전까지의 값 0.014594444444444441
실제값 : 0.014



---
## 의문점 쓰는 공간 

1. view_rate의 마지막 행과 열은 각각 무엇을 의미할까요???
> 마지막 열 : 93, 120, 628, 652, 745, 785, 837, 962, 988, 1206 위치에 있는 것들을 제외하고는 평균의 소수점 넷째자리에서 반올림한 값입니다. 
> <br>위의 예외들은 올린것도 있고...버린것도있고... 뒤죽박죽 
> <br>마지막 행 : 위와 마찬가지 ;;; 
